### ISSUE TITLE
no-unused-vars: False positive for recursive arrow function

### ISSUE THREAD

**Issue by JHilker** [eslint/eslint#12572]
**Tell us about your environment**

```
Environment Info:

Node version: v12.10.0
npm version: v6.10.3
Local ESLint version: v6.6.0
Global ESLint version: Not found
```

**What parser (default, Babel-ESLint, etc.) are you using?**

Occurs with both default and `babel-eslint@10.0.3`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  rules: {
    'no-unused-vars': 'error'
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

Reproducible demo: https://eslint.org/demo#eyJ0ZXh0IjoiY29uc3QgcmVjdXJzaXZlRm9vID0gKG51bSkgPT4ge1xuICBpZiAobnVtIDwgMCkgcmV0dXJuO1xuICBjb25zb2xlLmxvZyhudW0pO1xuICByZWN1cnNpdmVGb28obnVtIC0gMSk7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXVudXNlZC12YXJzIjoyfSwiZW52Ijp7Im5vZGUiOnRydWUsImVzNiI6dHJ1ZX19fQ==

```js
const recursiveFoo = (num) => {
  if (num < 0) return;
  console.log(num);
  recursiveFoo(num - 1);
}
```

```bash
yarn eslint foo.js
```

**What did you expect to happen?**

`no-unused-vars` Should not flag `recursiveFoo` as unused, since it is used within the context of itself

**What actually happened? Please include the actual, raw output from ESLint.**

`no-unused-vars` flags `recursiveFoo` as unused

```
➜  some-repo git: ✗ yarn eslint foo.js
yarn run v1.17.3
$ /some-repo/node_modules/.bin/eslint foo.js

/some-repo/foo.js
  1:7  error  'recursiveFoo' is assigned a value but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command
```

**Are you willing to submit a pull request to fix this bug?**

With some insight on where to best make the fix, sure. I haven't dug into why this could be occurring

**Comment by JHilker** [comment]
Nevermind, looks like this was intentionally part of https://github.com/eslint/eslint/issues/10982 / https://github.com/eslint/eslint/pull/11032
